### PR TITLE
[Docs] [Tune] `ResultGrid` Docs and API reference

### PR DESCRIPTION
**Pull request by amogkam**
Improve docstring for `ResultGrid` and show API reference and docstring in Tune API section.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/result_grid.py [hunk 1]
```diff
@@ -15,30 +15,70 @@
...
+        df = result_grid.get_dataframe()(metric="loss", mode="min")
 
     Note that trials of all statuses are included in the final result grid.
     If a trial is not in terminated state, its latest result and checkpoint as
     seen by Tune will be provided.
+
+    One may wonder how is ResultGrid different than ExperimentAnalysis. ResultGrid is
+    supposed to succeed ExperimentAnalysis. However, it has not reached the same
+    feature parity yet. For interacting with an existing experiment, located at
+    ``local_dir``, one may do the following:
```
Comment (richardliaw) [inline python/ray/tune/result_grid.py:76]: Can you make this more active voice?

```suggestion
   ResultGrid will be the successor of the ExperimentAnalysis object
   but is not yet at feature parity. For interacting with an existing experiment, located at
    ``local_dir``, do the following:
```

File: python/ray/tune/result_grid.py [hunk 2]
```diff
@@ -15,30 +15,70 @@
...
+
+        # Get the best checkpoint corresponding to the best result.
         best_checkpoint = best_result.checkpoint
-        best_metric = best_result.metric
+
+        # Get a dataframe for the last reported results of all of the trials
+        df = result_grid.get_dataframe()
+
+        # Get a dataframe for the minimum loss seen for each trial
+        df = result_grid.get_dataframe()(metric="loss", mode="min")
```
Comment (krfricke) [inline python/ray/tune/result_grid.py:67]:
```suggestion
        df = result_grid.get_dataframe(metric="loss", mode="min")
```

File: python/ray/tune/result_grid.py [hunk 3]
```diff
@@ -15,30 +15,69 @@
...
 
     You can use it to inspect the trials run as well as obtaining the best result.
 
-    The constructor is a private API.
+    The constructor is a private API. This object can only be created as a result of
+    ``Tuner.fit()``.
 
     Usage pattern:
 
     .. code-block:: python
```
Comment (Yard1) [inline python/ray/tune/result_grid.py:27]: can we test the docstrings here?

File: python/ray/tune/result_grid.py [hunk 4]
```diff
@@ -15,30 +15,69 @@
...
+        df = result_grid.get_dataframe()
+
+        # Get a dataframe for the minimum loss seen for each trial
+        df = result_grid.get_dataframe(metric="loss", mode="min")
 
     Note that trials of all statuses are included in the final result grid.
     If a trial is not in terminated state, its latest result and checkpoint as
     seen by Tune will be provided.
+
+   ResultGrid will be the successor of the ExperimentAnalysis object
```
Comment (Yard1) [inline python/ray/tune/result_grid.py:73]:
```suggestion
   ``ResultGrid`` will be the successor of the ExperimentAnalysis object
```

File: python/ray/tune/result_grid.py [hunk 5]
```diff
@@ -15,30 +15,69 @@
...
             result = result_grid[i]
             if not result.error:
                 print(f"Trial finishes successfully with metric {result.metric}.")
             else:
-                print(f"Trial errors out with {result.error}.")
-        best_result = result_grid.get_best_result()
+                print(f"Trial failed with error {result.error}.")
+
+
+    You can also use `result_grid` for more advanced analysis.
```
Comment (Yard1) [inline python/ray/tune/result_grid.py:53]:
```suggestion
    You can also use ``result_grid`` for more advanced analysis.
```

File: python/ray/tune/result_grid.py [hunk 6]
```diff
@@ -15,30 +15,69 @@
 
 @PublicAPI(stability="beta")
 class ResultGrid:
-    """A set of ``Result`` objects returned from a call to ``tuner.fit()``.
+    """A set of ``Result`` objects for interacting with Ray Tune results.
 
     You can use it to inspect the trials run as well as obtaining the best result.
```
Comment (Yard1) [inline python/ray/tune/result_grid.py:20]:
```suggestion
    You can use it to inspect the trials ran as well as obtaining the best result.
```

**Review by Yard1 — APPROVED** [review]
LGTM, some nits and a question

File: python/ray/tune/result_grid.py (hunk 5 above)
Comment (amogkam) [reply python/ray/tune/result_grid.py]: updated

File: python/ray/tune/result_grid.py (hunk 6 above)
Comment (amogkam) [reply python/ray/tune/result_grid.py]: updated

File: python/ray/tune/result_grid.py (hunk 4 above)
Comment (amogkam) [reply python/ray/tune/result_grid.py]: updated

File: python/ray/tune/result_grid.py (hunk 3 above)
Comment (amogkam) [reply python/ray/tune/result_grid.py]: Chatted with @krfricke offline. We can't actually test Ray code snippets in docstrings so we would have to skip everything anyways.

Also, Tune outputs won't really show up well when rendering the docstring.

Seems the best option right now is to leave as a python code block.

**Pull request merged**